### PR TITLE
postgresqlPackages.tsja: init at 0.5.0

### PR DESCRIPTION
**Pull request by chayleaf**
## Description of changes

This inits [tsja](https://www.amris.jp/tsja/index.html), a Postgresql extension for Japanese full-text search

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by chayleaf** [comment]
To test:

1. Add it to `services.postgresql.extraPlugins`
2. Run `psql <db name> < /run/current-system/sw/share/postgresql/extension/tsja.sql`
3. Open the repl and try parsing something: `SELECT * FROM ts_debug('japanese', 'PostgreSQLで日本語のテキスト検索ができます。');`

**Comment by chayleaf** [comment]
@auntieNeo @paveloom since this touches mecab

File: pkgs/servers/sql/postgresql/ext/tsja.nix [hunk 1]
```diff
@@ -0,0 +1,39 @@
+{ lib
+, stdenv
+, postgresql
+, mecab
+}:
```
Comment (paveloom) [inline pkgs/servers/sql/postgresql/ext/tsja.nix:5]:
```suggestion
{ lib
, stdenv
, fetchurl

, mecab
, postgresql
```

File: pkgs/servers/sql/postgresql/ext/tsja.nix [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+stdenv.mkDerivation rec {
+  pname = "tsja";
+  version = "0.5.0";
+
+  src = fetchTarball {
+    url = "https://www.amris.jp/tsja/tsja-${version}.tar.xz";
+    sha256 = "0hx4iygnqw1ay3nwrf3x2izflw4ip9i8i0yny26vivdz862m97w7";
+  };
```
Comment (paveloom) [inline pkgs/servers/sql/postgresql/ext/tsja.nix:14]:
```suggestion
  src = fetchurl {
    url = "https://www.amris.jp/tsja/tsja-${version}.tar.xz";
    hash = "sha256-Z8iwMQwS5mjnx8D8hgHB/5rxrTehrXhBRFcSUJzUAGI=";
  };
```

File: pkgs/servers/sql/postgresql/ext/tsja.nix [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+  postPatch = ''
+    substituteInPlace Makefile \
+      --replace /usr/local/pgsql ${postgresql} \
+      --replace -L/usr/local/lib "" \
+      --replace -I/usr/local/include ""
+    substituteInPlace tsja.c --replace /usr/local/lib/mecab ${mecab}/lib/mecab
+  '';
+
+  buildInputs = [ postgresql mecab ];
```
Comment (paveloom) [inline pkgs/servers/sql/postgresql/ext/tsja.nix:24]:
```suggestion
  buildInputs = [ mecab postgresql ];
```

**Review by paveloom — APPROVED** [review]

**Comment by chayleaf** [comment]
yes, good idea

File: pkgs/servers/sql/postgresql/ext/tsja.nix [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+}:
+
+stdenv.mkDerivation rec {
+  pname = "tsja";
+  version = "0.5.0";
+
+  src = fetchurl {
+    url = "https://www.amris.jp/tsja/tsja-${version}.tar.xz";
+    hash = "sha256-Z8iwMQwS5mjnx8D8hgHB/5rxrTehrXhBRFcSUJzUAGI=";
+  };
```
Comment (pbsds) [inline pkgs/servers/sql/postgresql/ext/tsja.nix:17]:
```suggestion
  src = fetchzip {
    url = "https://www.amris.jp/tsja/tsja-${version}.tar.xz";
    hash = "sha256-h59UhUG/7biN8NaDiGK6kXDqfhR9uMzt8CpwbJ+PpEM=";
  };
```

File: pkgs/servers/sql/postgresql/ext/tsja.nix [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  '';
+
+  passthru.tests.tsja = nixosTests.tsja;
+
+  meta = with lib; {
+    description = "PostgreSQL extension implementing Japanese text search";
+    homepage = "https://www.amris.jp/tsja/index.html";
+    maintainers = with maintainers; [ chayleaf ];
+    platforms = postgresql.meta.platforms;
+    license = licenses.postgresql;
```
Comment (pbsds) [inline pkgs/servers/sql/postgresql/ext/tsja.nix:42]: Not that acquainted with postgres packages, but the source tarball notes in `docs/COPYING`:

    Feml is copyrighted free software by KOYAMA Hiro <tac@amris.co.jp>.
    You can redistribute it and/or modify it under the terms of the GPL version 2 (see the file GPL).

File: nixos/tests/tsja.nix [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+            tsja
+          ];
+        };
+      };
+  };
+
+  testScript = ''
+    start_all()
+    master.wait_for_unit("postgresql")
+    master.sleep(10)
```
Comment (pbsds) [inline nixos/tests/tsja.nix:24]: why is this needed? Could you add a comment?

File: pkgs/servers/sql/postgresql/ext/tsja.nix [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+      --replace -I/usr/local/include ""
+    substituteInPlace tsja.c --replace /usr/local/lib/mecab ${mecab}/lib/mecab
+  '';
+
+  buildInputs = [ mecab postgresql ];
+
+  installPhase = ''
+    mkdir -p $out/lib $out/share/postgresql/extension
+    mv libtsja.so $out/lib
+    mv dbinit_libtsja.txt $out/share/postgresql/extension/libtsja_dbinit.sql
```
Comment (pbsds) [inline pkgs/servers/sql/postgresql/ext/tsja.nix:32]: Could you make this file executable? I'm not sure why, but all other postgres packages i look at mark these executable.

**Review by pbsds — CHANGES_REQUESTED** [review]
Result of `nixpkgs-review pr 262224` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>49 packages built:</summary>
  <ul>
    <li>groonga</li>
    <li>mecab</li>
    <li>memento</li>
    <li>postgresql11JitPackages.pgroonga</li>
    <li>postgresql11JitPackages.tsja</li>
    <li>postgresql11Packages.pgroonga</li>
    <li>postgresql11Packages.tsja</li>
    <li>postgresql12JitPackages.pgroonga</li>
    <li>postgresql12JitPackages.tsja</li>
    <li>postgresql12Packages.pgroonga</li>
    <li>postgresql12Packages.tsja</li>
    <li>postgresql13JitPackages.pgroonga</li>
    <li>postgresql13JitPackages.tsja</li>
    <li>postgresql13Packages.pgroonga</li>
    <li>postgresql13Packages.tsja</li>
    <li>postgresql14JitPackages.pgroonga</li>
    <li>postgresql14JitPackages.tsja</li>
    <li>postgresql14Packages.pgroonga</li>
    <li>postgresql14Packages.tsja</li>
    <li>postgresqlJitPackages.pgroonga (postgresql15JitPackages.pgroonga)</li>
    <li>postgresqlJitPackages.tsja (postgresql15JitPackages.tsja)</li>
    <li>postgresqlPackages.pgroonga (postgresql15Packages.pgroonga)</li>
    <li>postgresqlPackages.tsja (postgresql15Packages.tsja)</li>
    <li>postgresql16JitPackages.pgroonga</li>
    <li>postgresql16JitPackages.tsja</li>
    <li>postgresql16Packages.pgroonga</li>
    <li>postgresql16Packages.tsja</li>
    <li>python310Packages.fugashi</li>
    <li>python310Packages.fugashi.dist</li>
    <li>python310Packages.ipadic</li>
    <li>python310Packages.ipadic.dist</li>
    <li>python310Packages.mecab-python3</li>
    <li>python310Packages.mecab-python3.dist</li>
    <li>python310Packages.unidic</li>
    <li>python310Packages.unidic.dist</li>
    <li>python310Packages.wordfreq</li>
    <li>python310Packages.wordfreq.dist</li>
    <li>python311Packages.fugashi</li>
    <li>python311Packages.fugashi.dist</li>
    <li>python311Packages.ipadic</li>
    <li>python311Packages.ipadic.dist</li>
    <li>python311Packages.mecab-python3</li>
    <li>python311Packages.mecab-python3.dist</li>
    <li>python311Packages.unidic</li>
    <li>python311Packages.unidic.dist</li>
    <li>python311Packages.wordfreq</li>
    <li>python311Packages.wordfreq.dist</li>
    <li>tts</li>
    <li>tts.dist</li>
  </ul>
</details>

---

- [x] package path fits guidelines
- [x] package name fits guidelines
- [x] package version fits guidelines
- [x] package build on x86_64-linux
- [ ] executables tested on x86_64-linux
  - nixos test works though
- [x] `meta.description` is set and fits guidelines
- [x] `meta.license` fits upstream license
- [x] `meta.platforms` is set
- [x] `meta.maintainers` is set
- [x] build time only dependencies are declared in `nativeBuildInputs`
- [ ] source is fetched using the appropriate function
  - see suggestion
- [x] the list of `phases` is not overridden
- [x] when a phase (like `installPhase`) is overridden it starts with `runHook preInstall` and ends with `runHook postInstall`.
- [x] patches have a comment describing either the upstream URL or a reason why the patch wasn't upstreamed
- [x] patches that are remotely available are fetched rather than vendored

File: pkgs/servers/sql/postgresql/ext/tsja.nix (hunk 5 above)
Comment (chayleaf) [reply pkgs/servers/sql/postgresql/ext/tsja.nix]: thanks for catching this!

File: nixos/tests/tsja.nix (hunk 6 above)
Comment (chayleaf) [reply nixos/tests/tsja.nix]: I've seen other postgresql extension tests do this, but it seems to work without this line, so I removed it

File: pkgs/servers/sql/postgresql/ext/tsja.nix (hunk 7 above)
Comment (pbsds) [reply pkgs/servers/sql/postgresql/ext/tsja.nix:32]: Maybe not explicitly in the builders, but every package build I've inspected has executable `.sql` files. Try for example:

    fd '\.sql$' $(nix-build . -A postgresql14Packages.smlar --no-out-link) --type x

I've been search for why this might be the case, but I have not gotten any wiser. I don't consider this a blocker.

**Review by pbsds — DISMISSED** [review]

**Comment by pbsds** [comment]
Could you look into the darwin build failure? Likely it is missing a header, or you could mark it as broken

**Comment by chayleaf** [comment]
I've removed unsupported linker flags
